Step5 리뷰요청 드립니다!

aksgudwns opened the pull request:
안녕하세요 
step5 리뷰 요청드립니다.

앞에 커밋 여섯개(3c4536f ~ 32708ae)는 aksgudwns브랜치 로컬에서 작업하다만것을 실수로 push해서 되돌리는 작업을 했습니다.
커밋Id e17614e부터 리뷰 부탁드리겠습니다..

지난번에 다시 리뷰주신 부분도 반영했습니다
리스트도 같이 캐싱해야 했는데 제가 놓쳤습니다 추가 작업하였습니다..

Rank는 코드까지 친절하게 작성해주셔서 어떻게 반영하면 좋을지 많은 고민을 하였습니다..
말씀주신부분 작업 중에, Rank와 MatchBonus에 원시타입이 각각 따로 정의되어 있어, 유지보수에 어려울 거라는 생각이 들었습니다.
그래서 innerClass를 활용한 코드로 수정해보았는데, 의도하신대로 된 건지 모르겠습니다
이 부분도 같이 리뷰 부탁드립니다!
 
감사합니다.

aksgudwns committed:
refactor - for문 코드컨벤션 맞추어 수정, 문자파싱 책임분리 작업

aksgudwns committed:
doc - 보너스번호 요구사항 추가

aksgudwns committed:
feat - Rank enum추가 lottoNumber isBonus field추가

aksgudwns committed:
Merge branch 'aksgudwns' of https://github.com/aksgudwns/java-lotto-pro into aksgudwns

aksgudwns committed:
conflict방지를 위해 브랜치 동기화작업

aksgudwns committed:
충돌방지를 위한 커밋작업

aksgudwns committed:
doc - 수동로또 기능 요구사항 추가

aksgudwns committed:
feat - 수동로또 구매 로직 구현

- 수동으로 입력받은 로또 합치는 기능 구현
- 입력받은 숫자와 입력받은 로또 개수가 동일한지 확인
- 입력받은 숫자가 구매금액 이상인지 확인
-

aksgudwns committed:
feat - 화면작업추가, 수동구매 0개일때 기능 구현

aksgudwns committed:
refactor - 요구사항에 맞게 출력되도록 뷰 로직 수정, 추가된 기능 테스트 구현

aksgudwns committed:
refactor - Rank, LottoNumber 리뷰 반영

- lottoNumberMintoMax 리턴하는 리스트 캐싱작업 추가
- Rank, MatchBonus 리팩토링

dhmin5693 commented on README.md:
```diff
@@ -28,4 +28,9 @@ ex.) "aaa" => RuntimeException, "-1:9" => RuntimeException, "//;\n1.2.3" => Runt
 * 2등을 위해 추가 번호를 하나 더 추첨한다.
-* 당첨 통계에 2등도 추가해야 한다. 문구 : (5개 일치, 보너스 볼 일치(30000000원) - 0개)
\ No newline at end of file
+* 당첨 통계에 2등도 추가해야 한다. 문구 : (5개 일치, 보너스 볼 일치(30000000원) - 0개)
+* 수동으로 로또를 구매할 수 있다
+* 구매할 수 있는 숫자는 최대 구입금액을 넘을 수 없다
+* 수동으로 구매할 번호를 입력받는다
+* 수동으로 입력한 로또의 개수는 입력받은 숫자와 일치하여야 한다.
+* 수동으로 구매하는 로또 숫자가 0개인 경우 금액만큼 전부 자동으로 구매한다.
```
```suggestion
- [ ] 당첨 통계에 2등도 추가해야 한다. 문구 : (5개 일치, 보너스 볼 일치(30000000원) - 0개)
- [x] 수동으로 로또를 구매할 수 있다
- [x] 구매할 수 있는 숫자는 최대 구입금액을 넘을 수 없다
- [x] 수동으로 구매할 번호를 입력받는다
- [x] 수동으로 입력한 로또의 개수는 입력받은 숫자와 일치하여야 한다.
- [x] 수동으로 구매하는 로또 숫자가 0개인 경우 금액만큼 전부 자동으로 구매한다.
```

정리는 항상 잘해주시고 계시는데요,
`- [ ]` , `- [x]` 문법을 사용해보시면 진행 여부 점검에 도움이 되실거에요

aksgudwns replied:
정리를 더 꼼꼼히 해야하는데 항상 부족하다고 느낍니다ㅠㅠ
앞으로 진행여부 작성하는 습관 가져보겠습니다!

dhmin5693 commented on src/main/java/lotto/Application.java:
```diff
@@ -14,15 +18,47 @@ public class Application {
+                getBuyCountLotto(lottoInputView.readUserInput("수동으로 구매할 로또 수를 입력해 주세요."),
+                        buyAmountDto);
+
+        if(buyCountLottoDto.getDirectBuyCount() == 0) {
```
0 을 상수로 만들어주세요.

```text
규칙 3: 모든 원시값과 문자열을 포장한다.
```

aksgudwns replied:
상수로 만들겠습니다

dhmin5693 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -3,18 +3,20 @@
     private static final int LOTTO_MIN_NUMBER = 1;
     private static final int LOTTO_MAX_NUMBER = 45;
     private static Map<Integer,LottoNumber> lottoNumbers = new HashMap<>();
     private int lottoNumber;
+    private static List<LottoNumber> lottoNumberList = new ArrayList<>();
```
자바 컨벤션 중 클래스 구성요소 정렬에 대한 내용이 있어요.

```
1. 상수
2. 필드
3. 생성자
4. 메소드
```

상수는 필드 위에 위치시켜주세요 :)

```suggestion
    private static final List<LottoNumber> lottoNumberList = new ArrayList<>();
  
    private int lottoNumber;
```

한 칸 정도 띄워주시면 보기에 더 좋습니다.
그리고 Map과 List는 final로 고정시켜 주시는게 더 좋을 것 같습니다.
참조가 변하진 않을거라고 판단되거든요.

aksgudwns replied:
상수로 선언하고 다시 작성하였습니다..!
상수랑 필드사이에 한 줄 띄우니 더 읽기 좋아진 것 같습니다!

dhmin5693 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -1,63 +1,103 @@
+    }
+}
+
+class MatchBonus {
```
방향은 굉장히 좋았다고 생각하는데요,
inner class라기보다는 한 파일에 클래스가 두 개 작성되어 있어요.

inner class는 개인적으로 선호하진 않다보니 저라면 별도의 파일로 분리해둘 것 같습니다.

aksgudwns replied:
네.. 파일분리하겠습니다!

dhmin5693 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -1,63 +1,103 @@
 
     public static Rank valueOf(int countOfMatch, boolean matchBonus) {
-        return RANK_BY_MATCH_COUNT.get(MatchBonus.of(countOfMatch,matchBonus));
+        Rank rank = RANK_BY_MATCH_COUNT.get(MatchBonus.of(countOfMatch, matchBonus));
+        return rank == null ? MISS : rank;
```
```suggestion
        return RANK_BY_MATCH_COUNT.getOrDefault(MatchBonus.of(countOfMatch, matchBonus), MISS);
```

getOrDefault를 활용해보세요.

aksgudwns replied:
Map쓸 때 getOrDefault 항상 염두하는 습관 들이겠습니다!

dhmin5693 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -1,63 +1,103 @@
+    final static MatchBonus MATCH_COUNT_MISS = new MatchBonus(0, false);
+
+    final static Map<Integer, MatchBonus> MATCH_COUNT_BY_MATCH_BONUS_FALSE = new HashMap<Integer, MatchBonus>() {
+        {
+            put(6, new MatchBonus(6, false));
+            put(5, new MatchBonus(5, false));
+            put(4, new MatchBonus(4, false));
+            put(3, new MatchBonus(3, false));
+        }
```
https://www.baeldung.com/java-double-brace-initialization
이중 중괄호 초기화는 요샌 권장되지 않는 방법이기는 합니다.
참고로만 알아주세요.

aksgudwns replied:
메모리누수가 발생할 수도 있다는 사실은 몰랐네요..!
실무에서는 사용하지 말아야겠습니다!
좋은 참고자료 공유해주셔서 감사합니다:)

dhmin5693 replied:
깔끔하게 제거해주셨네요 👍

dhmin5693 commented on src/main/java/lotto/dto/LotteriesDto.java:
```diff
@@ -21,4 +23,24 @@ public Lotteries getLotteriesDomain() {
+        return autoBuyCount;
+    }
+
+    public void setDirectBuyCount(int directBuyCount) {
+        this.directBuyCount = directBuyCount;
+    }
```
dto여도 setter는 지양해주세요.
삭제 부탁드리겠습니다.

aksgudwns replied:
삭제하겠습니다!

dhmin5693 reviewed (changes requested):
안녕하세요 형준님!

과제를 열심히 수행해주시고 피드백을 잘 반영해주시다보니
이번 미션에는 큰 피드백을 드릴만한 부분이 없어보입니다.

마지막 미션이니 마무리를 잘 해보자라는 느낌으로 몇 가지 의견 남겨보았으니 확인 부탁드립니다.
곧 TDD 미션이 끝나겠네요.
조금만 더 힘내주세요 💪

aksgudwns committed:
doc - [ ], [x] 를 이용한 진행여부 작성 추가

aksgudwns committed:
refactor - MatchBonus 파일 분리

aksgudwns committed:
refactor - 코드컨벤션 맞추는 작업

- 모든 원시값과 문자열 포장
- 클래스 구성요소 정렬순서 맞춤
- setter 삭제
- LottoNumber, MatchBonus final 접근자 추가

aksgudwns commented:
동휘님 그동안 세세히 리뷰해주셔서 감사합니다!
가이드 잘해주신 덕분에 계속 생각하게 되고, TDD, DDD에 대해 조금씩 알아가는 느낌을 받았습니다:)

리뷰해주신 부분 작업해보았습니다.
작업 중 다른 클래스에도 코드컨벤션이 안지켜진 부분이 있어 같이 수정하였습니다.
LottoNumber와 MatchBonus는 인스턴스가 상수처럼 쓰이는 클래스여서, 
필드들은 생성 후 변경될일이 없다고 판단하여 final접근자를 추가하였습니다.

혹시 제가 잘못 반영한 부분이 있다면 피드백 부탁드립니다!
감사합니다.

dhmin5693 commented on src/main/java/lotto/domain/MatchBonus.java:
```diff
@@ -1,33 +1,48 @@
-            matchBonusFalses.put(countOfMatch, new MatchBonus(countOfMatch, false));    
+
+    static final MatchBonus MATCH_COUNT_MISS = new MatchBonus(0, false);
+    static final Map<Integer, MatchBonus> MATCH_COUNT_BY_MATCH_BONUS_FALSE = new HashMap<Integer, MatchBonus>() {
+        {
+            put(6, new MatchBonus(6, false));
+            put(5, new MatchBonus(5, false));
+            put(4, new MatchBonus(4, false));
+            put(3, new MatchBonus(3, false));
         }
```
이 부분은 저번에 일부러 남기지 않은 부분이긴 한데, 어짜피 끝난 마당에 고민해보셨으면 하여 작성해봅니다.

이중 괄호 초기화 문법을 사용하지 않고도 final 상태를 유지할 수 있습니다.
stream을 적절히 활용하면 되겠네요.
방법이야 많겠지만 한 가지 힌트를 드리자면 map 단계에서 entry를 생성하고 넘길 수도 있습니다.

aksgudwns replied:
마지막까지 코멘트 주셔서 감사합니다!
말씀주신 내용 고민해보고 다음 과제부터 반영해보겠습니다!

dhmin5693 reviewed (approved):
피드백 반영 잘해주셨습니다.
TDD 진행하느라 고생 너무 많으셨어요.
이번 미션을 이만 마무리해도 될 것 같습니다.

첫 번째 미션부터 쉽지 않아서 꽤 고되셨을텐데 첫 고개를 훌륭히 넘어가셨네요.
형준님의 남은 여정도 응원합니다.
이번 과정을 통해 많은 인사이트 얻어가시기 바랍니다.
